src/interchange/src/avro/encode.rs

```diff
@@ -150,30 +144,87 @@ impl DocTarget {
     }
 }
 
-/// Generates key and value Avro schemas
-pub struct AvroSchemaGenerator {
-    value_columns: Vec<(ColumnName, ColumnType)>,
-    key_info: Option<KeyInfo>,
-    writer_schema: Schema,
+/// Generates an Avro schema for a key type
+pub struct AvroKeySchemaGenerator {
+    columns: Vec<(ColumnName, ColumnType)>,
+    schema: Schema,
+}
+
+impl fmt::Debug for AvroKeySchemaGenerator {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("SchemaGenerator")
+            .field("writer_schema", &self.schema())
+            .finish()
+    }
+}
+
+impl AvroKeySchemaGenerator {
+    pub fn new(
+        desc: RelationDesc,
+        AvroSchemaOptions {
+            avro_key_fullname,
```

Comment:
This struct feels denormalized which causes the duplication of structs in this module. In principle this module doesn't care if it's a key or a value schema. If there is a distinction for external reasons then the burden should be placed on the calling crate to pass the right arguments. We can either require a flat list of arguments like like `build_row_schema_json` or we could define an `mz_interchange`-specific `AvroSchemaOptions` struct which is agnostic to the purpose of the schema.

A more general question is: does `AvroSchemaOptions` need to be denormalized in the first place? Can't we store one instance of it for the key and one for the value?

Reply:
Yes you're right, this is definitely a leaky abstraction making the `mz_interchange` logic more complicated.

> If there is a distinction for external reasons then the burden should be placed on the calling crate to pass the right arguments. We can either require a flat list of arguments 

Yep I'll do this for the interchange interface, and I'll see how it looks to refactor the `AvroSchemaOptions` in parsing and whether that ends up being cleaner or not